lib/vrl/vrl/src/lib.rs

```diff
@@ -10,40 +10,47 @@
 pub mod prelude;
 mod runtime;
 
+use compiler::Compiler;
 pub use compiler::{
     function, state, value, Context, Expression, Function, MetadataTarget, Program, ProgramInfo,
     SecretTarget, Target, TargetValue, TargetValueRef, VrlRuntime,
 };
 pub use diagnostic;
+use lookup::LookupBuf;
 pub use runtime::{Runtime, RuntimeResult, Terminate};
 pub use vector_common::TimeZone;
 
 /// Compile a given source into the final [`Program`].
 pub fn compile(source: &str, fns: &[Box<dyn Function>]) -> compiler::Result {
     let mut state = state::ExternalEnv::default();
 
-    compile_with_state(source, fns, &mut state)
+    compile_with_external(source, fns, &mut state)
 }
 
-pub fn compile_with_state(
+pub fn compile_with_external(
     source: &str,
     fns: &[Box<dyn Function>],
-    state: &mut state::ExternalEnv,
+    external: &mut state::ExternalEnv,
 ) -> compiler::Result {
-    let ast = parser::parse(source)
-        .map_err(|err| diagnostic::DiagnosticList::from(vec![Box::new(err) as Box<_>]))?;
-
-    compiler::compile_with_state(ast, fns, state)
+    compile_with_state(source, fns, external, state::LocalEnv::default())
 }
 
-pub fn compile_for_repl(
+pub fn compile_with_state(
     source: &str,
     fns: &[Box<dyn Function>],
     external: &mut state::ExternalEnv,
     local: state::LocalEnv,
-) -> compiler::Result<Program> {
+) -> compiler::Result {
     let ast = parser::parse(source)
         .map_err(|err| diagnostic::DiagnosticList::from(vec![Box::new(err) as Box<_>]))?;
 
-    compiler::compile_for_repl(ast, fns, local, external)
+    // Prevent mutating anything under the "vector" path in metadata. There are no cases
+    // where this should be allowed in VRL.
+    //
+    // This path is used to differentiate between log namespaces. It also contains
+    // metadata that transforms / sinks may rely on, so setting it to read-only
+    // prevents users from potentially breaking behavior relying on it.
+    external.add_read_only_metadata_path(LookupBuf::from("vector"), true);
```

Comment:
Right now there are still 3 "external" functions VRL exposes for compiling code. `compile`, `compile_with_external`, and `compile_with_state`. If this were moved outside of VRL, it would need to be added to every place that calls any of those 3 functions. I considered cleaning this up more to expose only a single `compile`, but I didn't want to make this PR even larger than it already is.

Given that, and we currently have no plans to ever separate VRL from Vector, I don't see a great reason to move this. If for some reason we change our mind and need a clean separation between Vector and VRL I'd be glad to revisit this.

Reply:
I respectfully disagree with the conclusion. This is a slippery slope, and one we've gone down before, and had to invest time to claw back a clear separation between the VRL API, and Vector.

> we currently have no plans to ever separate VRL from Vector

That's actually not true, we have been discussing this, and in general still think it's a good step to take, even though we haven't fleshed out the details yet (including some potential downsides we'd have to address first).

My proposal would be to either (a) add `external.add_read_only_metadata_path(LookupBuf::from("vector"), true)` to all places outside vector where we call `compile_with_external`, or (b) wrap the three functions you mentioned in a module inside the Vector codebase, and call those instead of directly calling the VRL ones. It's really not that big of a task to make this happen, so I'd like to see that done before we merge this PR.

Reply:
> My proposal would be to [...] wrap the three functions you mentioned in a module inside the Vector codebase, and call those instead of directly calling the VRL ones. It's really not that big of a task to make this happen, so I'd like to see that done before we merge this PR.

That sounds fairly straight-forward to me

Reply:
I had a chat with Nathan about this today. He made the good point that, even if we do extract it, which we don't have imminent plans for, we don't yet know the boundaries without additional use-cases for VRL outside of Vector. We could defer until we get to the point of extraction to figure out the right boundaries rather than trying to anticipate what they might be.

A counter point to this would be if there is any value to this conceptual separation if VRL never leaves Vector. Like does it make it more maintainable, less bug-prone, etc..

My thought would be that we leave it as-is and loop this discussion of what belongs into VRL into the broader conversations around VRL's future that we are planning to develop some guidelines.

Reply:
> A counter point to this would be if there is any value to this conceptual separation if VRL never leaves Vector. Like does it make it more maintainable, less bug-prone, etc..

That is precisely the reason why I've been focused on keeping those boundaries more clear. Yes, it'll make separating VRL easier, but it also makes it easier to reason about Vector and VRL separately, and create a clear API boundary between the two.

For those reasons, I'd still like to see changes made to this part of the PR, it doesn't even have to be a lot of work (I suspect this discussion took longer than the work needed to create that boundary). I'm not advocating for a specific change, but I am advocating for a solution that pulls this change out of the vrl "workspace".

The solution _might_ even be to just move the entire `vrl` crate (which is more of a "let's group every piece of VRL together in one package for easy access" crate) out of `lib/vrl` and have it live as `lib/vector-vrl` for example. That makes it clear that this crate exists as the link between Vector and VRL, and isn't specific to VRL itself.

> That means it's pulling in the functions from vector-vrl-functions, and needs to keep the read-only properties that will exist in remap.

That's also why we created a separate crate in the first place, it makes it clear that those functions are Vector specific, and the VRL CLI is something we built for Vector specifically (it's still more a hobby project grown out of proportions than anything else, so I wouldn't want to use this as a reason for any changes in the VRL compiler).